### ISSUE TITLE
epic: Customer Tasks Page

### ISSUE THREAD

**Issue by davidtaing** [codesydney/migram-frontend#133]
I've found the complexity of this task to be really high, so I'll have to split this into smaller chunks.

I'll outline my research and decisions in the comments below.

Related: #132 

### Components

- [x] Tasks Page
- [ ] ~~Tasks List~~ _Removed_
- [x] Task Card

### APIs

- [ ] ~~Fetch Tasks~~
- [ ] ~~Fetch Offers Of Tasks~~
- [ ] ~~Accept Offer~~
- [ ] ~~View Provider Contact Details~~

## Approach
- Start off with a skeleton task page
- Work on the Task Card component
- Then complete the Tasks Page / List components

## Edit 10th Feb 2023

This feature is nearly complete. We'll need to add some sort of notification when a request is made. And a small feature to lookup a the provider's details.

- [ ] #141

**Comment by davidtaing** [comment]
# Task Card
**Initial View**
![image](https://user-images.githubusercontent.com/8443215/217207316-16cb358f-2ee1-4888-9bc5-c683c37221ec.png)

**After "View Offers" button has been clicked**
![image](https://user-images.githubusercontent.com/8443215/217207150-7c8246e8-0f6e-4d5c-898d-788fb473e61e.png)

The TaskCard is main component of the page.

Users should be able to click on the view

**Comment by davidtaing** [comment]
# Fetch Offers of Tasks - API

"Get all the offers that belong to this task"

That one was an interesting one and required a good amount of research.

I suspect this would be a high-traffic query, so I've looked into ways that we can improve performance.

## Database Schema Overview

![Schema Overview](https://user-images.githubusercontent.com/8443215/217222406-51272032-9270-4134-94fd-e1504333140c.png)

## Current Stance

My current stance is that we should wait until we have traffic, then use analytics to determine what queries to optimize.

## Potential Options and Their Tradeoffs

### 1. Non-Indexed Document Reference on the Many Collection [Currently implemented]

**Pros**
- Less Complexity: Simplest to implement

**Cons**
- Poorer Performance: Requires a collection scan to lookup items - O(N) runtime

### 2. Multiple Document References on the One Collection
**Pros**
- Better Read Performance: Index lookups can be faster than collection scans - O(log N)

**Cons**
- Worse Write Performance: Operations will require writes to both Tasks and Offers collections.
- Additional Application Complexity: Due to having to maintain data in two locations and the con above.

### 3. Indexed Document References on the Many Collection
**Pros**
- Better Performance: Index lookups can be faster than collection scans - O(log N)
- Less Complexity: Does not require any changes to the one collection.

**Cons**
- Poorer Performance: Requires a collection scan to lookup items - O(N) runtime

## Conclusion

For my conclusion, I've factored in the other pages and the other queries such as:
- Customer Tasks Page - "Get me all tasks that belong to this `Customer User`",
- Provider Offers Page - "Get me all offers that belong to this `Provider User`"

I've made the assumption that these queries will have higher traffic than the one discussed here.

- Index on `Task.CustomerId` and `Offer.ProviderId`

For that reason, I would go with option 1 for the page queries instead. And go with option 2 for this Fetch Offers of Tasks Query 

- Multiple Offer references on the Tasks collection


## Other Comments

- It would make sense to implement a new `GET /tasks/:taskId/offers` endpoint. But I think it's best to use the existing endpoint at `GET /offers` for now.

**Comment by davidtaing** [comment]
After our meeting with Amro and Engramar last night. I have decided to reuse some of the old code that handles the API requests.

I'll try to get the full signup and full task/offer flows unblocked, and I'll evaluate coming back and further cleaning up the feature.
